File: opcua-webservice-node/src/server/addressspace/nodes/UaVariableType.ts
import { Attributes, NodeClass } from "opcua-webapi";
import { UaLocalizedText, UaNodeId, UaVariant, UaVariantType, VariableIds, VariableTypeIds } from "opcua-webapi-ts";
import { UaDefinitionNode } from "./UaDefinitionNode";
import type { UaVariable } from "./UaVariable";

export class UaVariableType extends UaDefinitionNode {
    private readonly _dataType: UaNodeId;
    private readonly _valueRank: number;
    private _value: UaVariant;

    constructor(
        nodeId: UaNodeId,
        browseName: string,
        displayName: UaLocalizedText,
        isAbstract: boolean,
        dataType: UaNodeId,
        valueRank: number) {
        super(nodeId, browseName, displayName, isAbstract);
        this._dataType = dataType;
        this._valueRank = valueRank;
        this._value = UaVariant.null();
    }

    get nodeClass(): NodeClass {
        return NodeClass.VariableType;
    }

    get dataType(): UaNodeId {
        return this._dataType;
    }

    get valueRank(): number {
        return this._valueRank;
    }

    get value(): UaVariant {
        return this._value;
    }

    set value(value: UaVariant) {
        this._value = value;
    }

    addMember(member: UaVariable): void {
        this.addMemberNode(member);
    }

    getAttribute(attributeId: number): UaVariant {
        if (Attributes.Value === attributeId) {
            return this._value;
        }
        if (Attributes.DataType === attributeId) {
            return UaVariant.nodeId(this._dataType);
        }
        if (Attributes.ValueRank === attributeId) {
            return UaVariant.integer(this._valueRank, UaVariantType.Int32);
        }

        return super.getAttribute(attributeId);
    }
}

File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodes/UaVariableType.java
package org.opcfoundation.webserver.addressspace.nodes;

import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.springframework.lang.Nullable;

public class UaVariableType extends UaDefinitionNode {
    private final NodeId dataType;
    private final int valueRank;
    private Variant value;

    public UaVariableType(
            NodeId nodeId,
            String browseName,
            LocalizedText displayName,
            boolean isAbstract,
            NodeId dataType,
            int valueRank)
    {
        super(nodeId,browseName,displayName,isAbstract);
        this.dataType = dataType;
        this.valueRank = valueRank;
    }

    @Override
    public NodeClass nodeClass() {
        return NodeClass.VariableType;
    }

    public NodeId dataType() {
        return dataType;
    }

    public int valueRank() {
        return valueRank;
    }

    public Variant value() {
        return value;
    }

    public void setValue(Variant value)
    {
        this.value = value;
    }

    public void addMember(UaVariable member)
    {
        addMemberNode(member);
    }

    @Override
    public Variant getAttribute(int attributeId)
    {
        if (AttributeId.Value.id() == attributeId)
        {
            return value;
        } else if (AttributeId.DataType.id() == attributeId) {
            return new Variant(dataType);
        } else if (AttributeId.ValueRank.id() == attributeId) {
            return new Variant(valueRank);
        }

        return super.getAttribute(attributeId);
    }
}
